File: DAA/MinMax.py

def minMaxDivide(arr):
    arrayLen = int(len(arr))
    subArraySize = int(arrayLen/2)
    # 1,2,3,4,5
    # 1,2   3,4,5
    #       3,4   5
    #
    # 1,2,3,4
    # 1,2   3,4
    if subArraySize > 1:
        leftArr = arr[:subArraySize]
        rightArr = arr[subArraySize:]
        lMin, lMax = minMaxDivide(leftArr)
        rMin, rMax = minMaxDivide(rightArr)
        min = lMin
        max = lMax
        if min > rMin:
            min = rMin
        if max < rMax:
            max = rMax
        return min, max;
    if arrayLen == 3:
        #arr[0] = 3, arr[1]=4    arr[2] = 5
        min, max = minMaxDivide(arr[:2]) #[3,4]
        if arr[2] < min:
            return arr[2], max 
        elif arr[2] > max:
            return min, arr[2]
        else:
            return min, max

    if arr[0] < arr[1]:
       return arr[0], arr[1]
    return arr[1], arr[0]

if __name__ == "__main__":
    min,max = minMaxDivide([4,5,6,10,2,6,83,11,7,9])
    print("Min:",min)
    print("Max", max)

File: DAA/MinMax.cpp
#include <utility>
#include <vector>
#include <iostream>


std::pair<int, int> minMaxDivide(std::vector<int> a)
{
    int subArraySize = a.size()/2;
    if(subArraySize > 0){
        std::vector<int> subArrayLeft;
        subArrayLeft.assign(a.begin(), a.begin()+subArraySize);

        std::vector<int> subArrayRight;
        subArrayRight.assign(a.begin()+subArraySize,a.end());
        auto leftPair = minMaxDivide(subArrayLeft);
        auto rightPair= minMaxDivide(subArrayRight);
        int min = leftPair.first;
        int max = leftPair.second;
        if(min > rightPair.first)
            min = rightPair.first;
        if(max < rightPair.second)
            max = rightPair.second;
        return std::make_pair(min, max);
    }
    return std::make_pair(a[0], a[0]);
}


int main() 
{
    std::vector<int> list = {4,5,1,60,7,2,1,8,9,10,3};
    auto pair = minMaxDivide(list);
    std::cout<<"Min:"<< pair.first << " | Max: "<< pair.second << std::endl;
}
